source/extensions/clusters/aggregate/cluster_lb.cc

```diff
@@ -0,0 +1,121 @@
+      pair = linearizePrioritySet(clusters);
+  if (pair.first.hostSetsPerPriority().empty()) {
+    load_balancer_ = nullptr;
+  } else {
+    load_balancer_ =
+        std::make_unique<AggregateClusterLoadBalancer::LoadBalancerImpl>(*this, std::move(pair));
+  }
+}
+
+void AggregateClusterLoadBalancer::onClusterAddOrUpdate(Upstream::ThreadLocalCluster& cluster) {
+  if (std::find(clusters_.begin(), clusters_.end(), cluster.info()->name()) != clusters_.end()) {
+    refresh();
+
+    cluster.prioritySet().addPriorityUpdateCb(
+        [this](uint32_t, const Upstream::HostVector&, const Upstream::HostVector&) { refresh(); });
+    cluster.prioritySet().addMemberUpdateCb(
+        [this](const Upstream::HostVector&, const Upstream::HostVector&) { refresh(); });
+  }
+}
+
+void AggregateClusterLoadBalancer::onClusterRemoval(const std::string& cluster_name) {
+  auto clusters = clusters_;
+  auto it = std::find(clusters.begin(), clusters.end(), cluster_name);
+  if (it != clusters.end()) {
+    clusters.erase(it);
+    refresh(clusters);
+  }
+}
+
+std::pair<Upstream::PrioritySetImpl,
+          std::vector<std::pair<uint32_t, Upstream::ThreadLocalCluster*>>>
+AggregateClusterLoadBalancer::linearizePrioritySet(const std::vector<std::string>& clusters) {
+  Upstream::PrioritySetImpl priority_set;
+  std::vector<std::pair<uint32_t, Upstream::ThreadLocalCluster*>> priority_to_cluster;
+  int next_priority_after_linearing = 0;
+
+  // Linearize the priority set. e.g. for clusters [C_0, C_1, C_2] referred in aggregate cluster
+  //    C_0 [P_0, P_1, P_2]
+  //    C_1 [P_0, P_1]
+  //    C_2 [P_0, P_1, P_2, P_3]
+  // The linearization result is:
+  //    [C_0.P_0, C_0.P_1, C_0.P_2, C_1.P_0, C_1.P_1, C_2.P_0, C_2.P_1, C_2.P_2, C_2.P_3]
+  // and the traffic will be distributed among these priorities.
+  for (const auto& cluster : clusters) {
+    auto tlc = cluster_manager_.get(cluster);
+    if (tlc == nullptr) {
+      continue;
+    }
+
+    int priority_in_current_cluster = 0;
+    for (const auto& host_set : tlc->prioritySet().hostSetsPerPriority()) {
+      if (!host_set->hosts().empty()) {
+        priority_set.updateHosts(
+            next_priority_after_linearing++, Upstream::HostSetImpl::updateHostsParams(*host_set),
+            host_set->localityWeights(), host_set->hosts(), {}, host_set->overprovisioningFactor());
+        priority_to_cluster.emplace_back(std::make_pair(priority_in_current_cluster, tlc));
+      }
+      priority_in_current_cluster++;
+    }
+  }
+
+  return std::make_pair(std::move(priority_set), std::move(priority_to_cluster));
+}
+
+void AggregateClusterLoadBalancer::updatePrioritySetCallbacks(
+    const std::vector<std::string>& clusters, PriorityCb priority_cb, MemberCb member_cb) {
+  for (const auto& cluster : clusters) {
+    auto tlc = cluster_manager_.get(cluster);
+    if (tlc == nullptr) {
+      continue;
+    }
+
+    tlc->prioritySet().addPriorityUpdateCb(priority_cb);
```

Comment:
Will the priority change trigger member update cb or member change trigger priority update cb? I didn't check the code. If it's correct, we can remove one of them, otherwise, we might lose some update info if we only bind cb to one of them.

Reply:
the priority update cb is triggered whenever a HostSet is updated, and member update is triggered whenever there's a change  to *any* host set. The added/removed hosts passed alongside are added/removed to the hostset/overall cluster respectively.

Reply:
Removed priority cb